src/pages/workspaces/hooks/useDeletionPolling.ts

```diff
@@ -0,0 +1,137 @@
+  workspaceId: string,
+  state: WorkspaceState,
+  errorMessage?: string
+): Workspace[] =>
+  workspaces.map((ws) => {
+    if (ws.workspace.workspaceId === workspaceId) {
+      const update = _.cloneDeep(ws);
+      update.workspace.state = state;
+      update.workspace.errorMessage = errorMessage;
+      return update;
+    }
+    return ws;
+  });
+
+const updateWorkspace = (
+  workspace: Workspace,
+  workspaceId: string,
+  state: WorkspaceState,
+  errorMessage?: string
+): Workspace => {
+  if (workspace.workspace.workspaceId === workspaceId) {
+    const update = _.cloneDeep(workspace);
+    update.workspace.state = state;
+    update.workspace.errorMessage = errorMessage;
+    return update;
+  }
+  return workspace;
+};
+
+export const useDeletetionPolling = (workspaces: Workspace[]) => {
+  // we have to do the signal/abort manually instead of with useCancelable so that the it can be cleaned up in the
+  // this component's useEffect, instead of the useEffect in useCancelable
+  // const workspaces: Workspace[] = useStore<Workspace[]>(workspacesStore)
+  const [controller, setController] = useState(new window.AbortController());
+  const abort = () => {
+    controller.abort();
+    setController(new window.AbortController());
+  };
+
+  useEffect(() => {
+    const checkWorkspaceDeletion = async (workspace: Workspace) => {
+      const doUpdate = (state: WorkspaceState, errorMessage?: string) => {
+        const workspaceId = workspace.workspace.workspaceId;
+        abort();
+        workspacesStore.update((wsList) => updateWorkspacesList(wsList, workspaceId, state, errorMessage));
+        workspaceStore.update((ws) => updateWorkspace(ws, workspaceId, state, errorMessage));
+      };
+
+      try {
+        const wsResp: Workspace = await Ajax(controller.signal)
+          .Workspaces.workspace(workspace.workspace.namespace, workspace.workspace.name)
+          .details(['workspace.state', 'workspace.errorMessage']);
+        const state = wsResp.workspace.state;
+
+        if (state === 'DeleteFailed') {
+          doUpdate(state);
+        }
+      } catch (error) {
+        if (error instanceof Response && error.status === 404) {
+          doUpdate('Deleted');
+        }
+      }
+    };
+    const iterateDeletingWorkspaces = async () => {
+      const deletingWorkspaces = _.filter((ws) => ws?.workspace?.state === 'Deleting', workspaces);
+      for (const ws of deletingWorkspaces) {
+        await checkWorkspaceDeletion(ws);
+      }
+    };
+
+    pollWithCancellation(() => iterateDeletingWorkspaces(), 30000, false, controller.signal);
+    return () => {
+      abort();
+    };
+    //  eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [workspaces]); // adding the controller to deps causes a double fire of the effect
+};
+
+// we need a separate implementation of this, because using useDeletetionPolling with a list
+// containing a single workspace in the WorkspaceContainer causes a rendering loop
```

Comment:
That seems… odd. But can't you at least pull out the `doUpdate` function and share it between the 2 hooks to reduce code duplication?

Reply:
Good call: at first I was only updating one store, but now that I'm updating both, they're identical.

I also found the render loop odd, but I double checked, and it's still there.